cargo-quickinstall/src/main.rs

```diff
@@ -31,30 +31,170 @@ fn main() -> Result<(), Box<dyn std::error::Error + Send + Sync + 'static>> {
         .crate_name
         .ok_or(InstallError::MissingCrateNameArgument(args::USAGE))?;
-    let version = match options.version {
-        Some(version) => version,
-        None => get_latest_version(&crate_name)?,
+
+    let version = options.version;
+    let target = options.target;
+
+    let f = if options.no_binstall {
+        do_main_curl
+    } else {
+        do_main_binstall
     };
-    let target = match options.target {
+
+    f(
+        crate_name,
+        version,
+        target,
+        options.dry_run,
+        options.fallback,
+    )
+}
+
+fn do_main_curl(
+    crate_name: String,
+    version: Option<String>,
+    target: Option<String>,
+    dry_run: bool,
+    fallback: bool,
+) -> Result<(), Box<dyn std::error::Error + Send + Sync + 'static>> {
+    let target = match target {
         Some(target) => target,
         None => get_target_triple()?,
     };
 
+    let version = match version {
+        Some(version) => version,
+        None => get_latest_version(&crate_name)?,
+    };
+
     let crate_details = CrateDetails {
         crate_name,
         version,
         target,
     };
 
-    if options.dry_run {
-        let shell_cmd = do_dry_run(&crate_details);
+    if dry_run {
+        let shell_cmd = do_dry_run_curl(&crate_details);
         println!("{}", shell_cmd);
         return Ok(());
     }
 
     let stats_handle = report_stats_in_background(&crate_details);
-    install_crate(&crate_details, options.fallback)?;
+    install_crate_curl(&crate_details, fallback)?;
     stats_handle.join().unwrap();
 
     Ok(())
 }
+
+struct Crate {
+    name: String,
+    version: Option<String>,
+}
+
+impl Crate {
+    fn into_arg(self) -> String {
+        let mut arg = self.name;
+
+        if let Some(version) = self.version {
+            arg.push('@');
+            arg += version.as_str();
+        }
+
+        arg
+    }
+}
+
+fn do_main_binstall(
+    crate_name: String,
+    version: Option<String>,
+    target: Option<String>,
+    dry_run: bool,
+    fallback: bool,
+) -> Result<(), Box<dyn std::error::Error + Send + Sync + 'static>> {
+    let is_binstall_compatible = get_cargo_binstall_version()
+        .map(
+            |binstall_version| match binstall_version.splitn(3, '.').collect::<Vec<_>>()[..] {
+                [major, minor, _] => {
+                    // If >=1.0.0
+                    major != "0"
+                            // Or >=0.17.0
+                            || minor
+                                .parse::<u64>()
+                                .map(|minor| minor >= 17)
+                                .unwrap_or(false)
+                }
+                _ => false,
+            },
+        )
+        .unwrap_or(false);
+
+    if !is_binstall_compatible {
+        do_main_curl("cargo-binstall".to_string(), None, None, dry_run, fallback)?;
+
+        if !dry_run {
+            println!("Bootstrap cargo-binstall itself to update $CARGO_HOME/.crates.toml");
```

Comment:
```suggestion
            println!("Bootstrapping cargo-binstall with itself to make `cargo uninstall ` work properly");
```

I haven't tried this out locally yet. would the full list of println messages for the bootstrapping flow make sense to someone who has not used rust?

Reply:
> I haven't tried this out locally yet. would the full list of println messages for the bootstrapping flow make sense to someone who has not used rust?

Not sure, but I stilll think it's better to at least print something so that they know quick-install is not stuck and is still making progress.